tools/parallelize/parallelize.cpp

```diff
@@ -0,0 +1,203 @@
+                this_->commandRunning = SIZE_MAX;
+                if (parent->runningConcurrency.fetch_sub(1, std::memory_order_relaxed) == 1) {
+                    parent->cv.notify_all();
+                }
+            } else {
+                this_->schedule(parent->nextCommandToRun++);
+            }
+
+            parent->update_display();
+        }
+
+        void schedule(std::size_t command) {
+            commandRunning = command;
+            executive.begin_execution(nullptr, parent->commands[command].first.data(), 0, nullptr);
+            tpWait.wait_for(executive.get_wait_handle());
+        }
+    };
+
+    void update_display() noexcept {
+        const auto commandsSize = commands.size();
+        const auto next         = nextCommandToRun;
+        const auto running      = runningConcurrency.load(std::memory_order_relaxed);
+        assert(running <= next);
+        printf("%zu scheduled %zu completed %zu running\n", commandsSize, next - running, running);
+    }
+};
+
+inline void schedule_command(parallelizer& p, const std::wstring_view commandPrefix, const std::wstring& native) {
+    std::wstring toExecute;
+    toExecute.reserve(commandPrefix.size() + native.size() + 1);
+    toExecute.assign(commandPrefix);
+    toExecute.push_back(L' ');
+    toExecute.append(native);
+    p.add_command(std::move(toExecute));
+}
+
+extern "C" int wmain(int argc, wchar_t* argv[]) {
+    try {
+        if (argc < 3) {
+            puts("Usage: parallelize.exe commandPrefix pathRoots");
+            return 1;
+        }
+
+        parallelizer p;
+        std::wstring_view commandPrefix(argv[1]);
+        for (int idx = 2; idx < argc; ++idx) {
+            std::filesystem::path thisSpec(argv[idx]);
+            if (std::filesystem::is_regular_file(thisSpec)) {
+                schedule_command(p, commandPrefix, thisSpec.native());
+                continue;
+            }
+
+            for (const auto& dirEntry : std::filesystem::recursive_directory_iterator(std::move(thisSpec))) {
+                if (!dirEntry.is_regular_file()) {
+                    printf("Skipping non-regular-file %ls\n", dirEntry.path().c_str());
+                    continue;
+                }
+
+                schedule_command(p, commandPrefix, dirEntry.path().native());
+            }
+        }
+
+        p.wait_all();
+        auto& ran          = p.results();
+        const auto success = [](const std::pair<std::wstring, std::optional<execution_result>>& command) noexcept {
+            auto& result = command.second.value();
+            return result.exitCode == 0 && result.output.empty();
+        };
+
+        const std::size_t successes = std::count_if(ran.begin(), ran.end(), success);
+        printf("%zu commands ran successfully.\n", successes);
+        if (successes == ran.size()) {
+            return 0;
+        }
+
+        for (auto& command : ran) {
+            if (success(command)) {

```

Comment:
It looks like you might be better off doing a partition between success and failure, that way you wouldn't have to rescan.

Reply:
Hmmm I think the second scan is fairly irrelevant for performance here, and would change the resulting printed command order. As it is we print results in a roughly depth-first order and that seems desirable.

Reply:
`stable_partition` is stable.

Reply:
Sure, but it's more than 2 passes over the input.

Reply:
I'm mostly interested in increasing the clarity of the logic, instead of micro-optimizing. stable_partition followed by a pointer subtraction to see the successful count, and then iterating through the failed out, avoids having to do a loop with continue.

Reply:
I guess in that case I just don't think a partition + a pointer subtraction is clearer than count_if.

Reply:
Ok to leave as-is, I just like partitioning things.